### PR TITLE
[ADD] sale_service: Add additional information in SOL display name

### PR DESCRIPTION
**Pull request by hugocarlier**
- Impacted modules:
    - sale_service (new module - installed only if sale
    AND (project OR planning) are installed)
    - sale
    - sale_project

- Description: When looking for a SOL in services apps, the displayed
name is of the form "SO ref - SOL name - unit price (partner
ref/name)". The presence of some of those elements may vary:
    - unit price is only displayed if it allows to clarify an
    ambiguity (i.e. two SOL with identical name but different unit
    price in the same SO).
    - partner name is only used if the partner ref does not exist.

- Implementation: The behavior currently existing in sale needed to be
extended for services apps. This has been done with the following
changes:
    - sale:
        - sale.order.line: Current behavior for the name_get (display
        partner ref) has been extended (display parnter name if has no
        ref) and split a additional method that is extended in a new
        module called..
    - .. sale_service:
        - sale.order.line: The additional behavior is extended to
        display the unit price of the SOL when it avoids ambiguity. A
        context parameter is used to keep this behavior inside
        services apps.

task-3052812

### CODE REVIEW

**Review by auon-odoo — CHANGES_REQUESTED** [review]
Thanks a lot! Be brave :fire:

File: addons/sale_service/models/sale_order_line.py
Comment (huca-odoo) [reply addons/sale_service/models/sale_order_line.py]: I actually made a mistake in this part of the code. The last line (name_per_id[line.id] = '- '+name) should be outside the condition. If not, the unit price won't be added to the name if there is no other additional info to display in the name (which is not really possible with the current implementation but could be in the future). I've updated my code accordingly.

**Comment by huca-odoo** [comment]
> Thanks a lot! Be brave fire

Thank you for you review and feedback! I added two comments and updated my code ;)

File: addons/sale_service/models/sale_order_line.py
Comment (huca-odoo) [reply addons/sale_service/models/sale_order_line.py]: I have updated the code using this method :)

File: addons/sale_service/models/sale_order_line.py
Comment (auon-odoo) [inline addons/sale_service/models/sale_order_line.py:51]:
```suggestion
                name_per_id[line.id] = f'- {name}'
```
looks better to me, to be more consistent with above line.

**Review by auon-odoo — APPROVED** [review]
LGTM :)

File: addons/sale_service/models/sale_order_line.py
Comment (xavierbol) [reply addons/sale_service/models/sale_order_line.py:51]: I agree. 😇

**Review by xavierbol — COMMENTED** [review]
Thanks for your work, I left some comments. 🙂

File: addons/project/report/project_task_burndown_chart_report.py
Comment (huca-odoo) [reply addons/project/report/project_task_burndown_chart_report.py:11]: I've added a test for the function remove_domain_leaf. Some tests are still failing: particularly in the following cases:
1) a part of the domain is [..., '|', '|', leaf_to_remove, leaf_to_remove, ...].
2) '!' are involved in a part of the domain with leaves that need to be removed.
I can look into that next week ;)

**Review by huca-odoo — COMMENTED** [review]
Thank you for your feedback, I've updated the code and left a comment for the test that has issues that still need to be addressed.

File: addons/resource/models/utils.py
Comment (xavierbol) [inline addons/resource/models/utils.py:1]:
```suggestion
# Part of Odoo. See LICENSE file for full copyright and licensing details.

from odoo.osv.expression import normalize_domain, NOT_OPERATOR, DOMAIN_OPERATORS
```

File: addons/resource/models/utils.py
Comment (xavierbol) [inline addons/resource/models/utils.py:7]:
```suggestion

```
There are some trailing spaces 🙂 
That is why the ci/style is not happy

Perhaps your editor can automatically remove the trailing spaces once you save? I can do that with mine. 🙂

File: addons/resource/models/utils.py
Comment (xavierbol) [inline addons/resource/models/utils.py:11]:
```suggestion

```
trailing spaces to remove. 🙂

File: addons/resource/models/utils.py
Comment (xavierbol) [inline addons/resource/models/utils.py:37]:
```suggestion

```

Trailing spaces to remove. 🙂

File: addons/resource/tests/__init__.py
Comment (xavierbol) [inline addons/resource/tests/__init__.py:1]:
```suggestion
# -*- coding: utf-8 -*-
# Part of Odoo. See LICENSE file for full copyright and licensing details.
```
😇

File: addons/resource/tests/test_utils.py
Comment (xavierbol) [inline addons/resource/tests/test_utils.py:6]: unused. 🙂

File: addons/sale_service/models/__init__.py
Comment (xavierbol) [inline addons/sale_service/models/__init__.py:1]:
```suggestion
# -*- coding: utf-8 -*-
# Part of Odoo. See LICENSE file for full copyright and licensing details.
```
😇 
In general, I try to put it in every python files. 🙂

File: addons/resource/models/utils.py
Comment (xavierbol) [inline addons/resource/models/utils.py:3]: In burndown chart, we could also use that method. 🙂 
(and so you can remove the function `remove_domain_leaf` in burndown chart too). 🙂

**Review by xavierbol — COMMENTED** [review]
Quick review. 🙂 

The ci/style failed because of unused imports and trailing spaces (in most of the cases). 🙁

File: addons/sale_service/models/__init__.py
Comment (xavierbol) [reply addons/sale_service/models/__init__.py:1]: Sorry but it is not done. 😬

Comment (huca-odoo) [reply addons/sale_service/models/__init__.py:1]: Ooops.. It is done now :+1:

**Review by xavierbol — COMMENTED** [review]
I just left one comment in this file `addons/sale_service/models/__init__.py`, after that it will be good. 🙂

**Review by xavierbol — APPROVED** [review]
LGTM, thanks again for your work. 🙂

File: addons/sale_project/models/sale_order_line.py
Comment (tivisse) [inline addons/sale_project/models/sale_order_line.py:22]: Probably needs a migration script to avoid a field recomputation on all the SO lines.

**Comment by xavierbol** [comment]
We no longer need a review from the Framework Python since the code no longer changes the base module. 🙂

**Comment by xavierbol** [comment]
We no longer need a review from the Security Team for the same reason. 🙂

**Review by xavierbol — APPROVED** [review]
LGTM, thanks again for your work. 🙂 

robodoo r+

**Comment by xavierbol** [comment]
robodoo rebase-ff

File: addons/sale_service/models/sale_order_line.py
Comment (KangOl) [inline addons/sale_service/models/sale_order_line.py:15]: The question is: why is this field stored anyway?

Comment (xavierbol) [reply addons/sale_service/models/sale_order_line.py:15]: @KangOl To be honest, I don't really know it was the case since 13.0 (and perhaps even before). 😬 
The field has been stored by that [commit](https://github.com/odoo/odoo/commit/7c1a1d15eebc50420e376138f45064059242640e) in the following PR: #18228. 😬 

Maybe to improve a little bit the performance since the field is used in many domains in the Project app? 🤔 🤷🏽‍♂️

Comment (KangOl) [reply addons/sale_service/models/sale_order_line.py:15]: Then, just add a `product_type` related field and use the domain `[('product_type', '=', 'service')]`. Or just add a search method on the field.

But this is for another task.

Comment (xavierbol) [reply addons/sale_service/models/sale_order_line.py:15]: Good idea, we will do that in another task. Thanks for your review 😁!

**Pull request closed without merging**
